akka-persistence-testkit/src/main/scala/akka/persistence/testkit/scaladsl/ImMemStorage.scala

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.testkit.scaladsl
+
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.{ BiFunction, Consumer }
+
+import akka.actor.{ ExtendedActorSystem, Extension, ExtensionId, ExtensionIdProvider }
+import akka.persistence.testkit.scaladsl.InMemStorageEmulator.JournalOperation
+import akka.persistence.testkit.scaladsl.ProcessingPolicy._
+import akka.persistence.testkit.scaladsl.SnapShotStorageEmulator._
+import akka.persistence.{ PersistentRepr, SelectedSnapshot, SnapshotMetadata, SnapshotSelectionCriteria }
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.util.{ Failure, Success, Try }
+
+trait InMemStorage[K, T] {
+
+  private val parallelism = Runtime.getRuntime.availableProcessors()
+
+  private final val eventsMap: ConcurrentHashMap[K, Vector[T]] = new ConcurrentHashMap()
```

Comment:
I see that you made things thread safe. Is that because this is an extension that is accessed both from the test itself and the journal actor?

An alternative would be to only access it via the journal actor by sending messages to it. In the `AsyncWriteJournal` implementation you can override `def receivePluginInternal: Actor.Receive` to handle such custom messages. Getting the ActorRef of the journal can be done with `Persistence(system).journalFor`. In that way the test methods (e.g. `failNextNPersisted`) could send messages to the journal actor.

Reply:
> I see that you made things thread safe. Is that because this is an extension that is accessed both from the test itself and the journal actor?
Yes, the storage is accessed both from the test and from the journal

The reason I didn't use `receivePluginInternal` is because you then must expose Futures in the methods to the users or make Awaits. I decided not to use Futures because other akka testkits do not have api methods with Futures (expectNext etc.) and avoiding them is much more convenient for the users.

Reply:
ok